doc/source/config.rst
Configuration Parameters
========================

This tutorial demonstrates how to set and get the TileDB
config parameters, and summarizes all current config parameters
explaining their function.

.. toggle-header::
    :header: **Example Code Listing**

    .. content-tabs::

       .. tab-container:: cpp
          :title: C++

          .. literalinclude:: {source_examples_path}/cpp_api/config.cc
             :language: c++
             :linenos:

You can create a config object and pass it to either a TileDB
context or VFS object as follows:

.. content-tabs::

   .. tab-container:: cpp
      :title: C++

      .. code-block:: c++

        // Create config object
        Config config;

        // Set/Get config to/from ctx
        Context ctx(config);
        Config config_ctx = ctx.config();

        // Set/Get config to/from VFS
        VFS vfs(ctx, config);
        Config config_vfs = vfs.config();

Compiling and running the program example of this tutorial
we get the output shown below. In the rest of the tutorial
we will discuss the various ways we used the config objects
in this program and explain the output.


.. code-block:: bash

   $ g++ -std=c++11 config.cc -o config_cpp -ltiledb
   $ ./config_cpp
   Tile cache size: 10000000

   Default settings:
   "sm.array_schema_cache_size" : "10000000"
   "sm.check_coord_dups" : "true"
   "sm.dedup_coords" : "false"
   "sm.enable_signal_handlers" : "true"
   "sm.fragment_metadata_cache_size" : "10000000"
   "sm.num_async_threads" : "1"
   "sm.num_tbb_threads" : "-1"
   "sm.tile_cache_size" : "10000000"
   "vfs.file.max_parallel_ops" : "8"
   "vfs.hdfs.kerb_ticket_cache_path" : ""
   "vfs.hdfs.name_node_uri" : ""
   "vfs.hdfs.username" : ""
   "vfs.min_parallel_size" : "10485760"
   "vfs.num_threads" : "8"
   "vfs.s3.connect_max_tries" : "5"
   "vfs.s3.connect_scale_factor" : "25"
   "vfs.s3.connect_timeout_ms" : "3000"
   "vfs.s3.endpoint_override" : ""
   "vfs.s3.max_parallel_ops" : "8"
   "vfs.s3.multipart_part_size" : "5242880"
   "vfs.s3.proxy_host" : ""
   "vfs.s3.proxy_password" : ""
   "vfs.s3.proxy_port" : "0"
   "vfs.s3.proxy_scheme" : "https"
   "vfs.s3.proxy_username" : ""
   "vfs.s3.region" : "us-east-1"
   "vfs.s3.request_timeout_ms" : "3000"
   "vfs.s3.scheme" : "https"
   "vfs.s3.use_virtual_addressing" : "true"

   VFS S3 settings:
   "connect_max_tries" : "5"
   "connect_scale_factor" : "25"
   "connect_timeout_ms" : "3000"
   "endpoint_override" : ""
   "max_parallel_ops" : "8"
   "multipart_part_size" : "5242880"
   "proxy_host" : ""
   "proxy_password" : ""
   "proxy_port" : "0"
   "proxy_scheme" : "https"
   "proxy_username" : ""
   "region" : "us-east-1"
   "request_timeout_ms" : "3000"
   "scheme" : "https"
   "use_virtual_addressing" : "true"

   Tile cache size after loading from file: 0


Setting/Getting config parameters
---------------------------------

*The TileDB config object is a simplified, in-memory key-value store/map,
which accepts only string keys and values*. The code below simply sets two parameters
and gets the value of a third parameter. We explain the TileDB parameters
at the end of this tutorial.

.. content-tabs::

   .. tab-container:: cpp
      :title: C++

      .. code-block:: c++

        Config config;

        // Set value
        config["vfs.s3.connect_timeout_ms"] = 5000;

        // Append parameter segments with successive []
        config["vfs."]["s3."]["endpoint_override"] = "localhost:8888";

        // Get value
        std::string tile_cache_size = config["sm.tile_cache_size"];
        std::cout << "Tile cache size: " << tile_cache_size << "\n\n";

The above code snippet produces the following output in our program:

.. code-block:: bash

   Tile cache size: 10000000


Iterating over config parameters
--------------------------------

TileDB allows you to iterate over the configuration parameters as well.
The code below prints the default parameters of a config object, as
we iterate before setting any new parameter value.

.. content-tabs::

   .. tab-container:: cpp
      :title: C++

      .. code-block:: c++

       Config config;
       std::cout << "Default settings:\n";
       for (auto& p : config) {
         std::cout << "\"" << p.first << "\" : \"" << p.second << "\"\n";
       }

The corresponding output is (note that we ran this on a machine with
8 cores):

.. code-block:: bash

   Default settings:
   "sm.array_schema_cache_size" : "10000000"
   "sm.check_coord_dups" : "true"
   "sm.dedup_coords" : "false"
   "sm.enable_signal_handlers" : "true"
   "sm.fragment_metadata_cache_size" : "10000000"
   "sm.num_async_threads" : "1"
   "sm.num_tbb_threads" : "-1"
   "sm.tile_cache_size" : "10000000"
   "vfs.file.max_parallel_ops" : "8"
   "vfs.hdfs.kerb_ticket_cache_path" : ""
   "vfs.hdfs.name_node_uri" : ""
   "vfs.hdfs.username" : ""
   "vfs.min_parallel_size" : "10485760"
   "vfs.num_threads" : "8"
   "vfs.s3.connect_max_tries" : "5"
   "vfs.s3.connect_scale_factor" : "25"
   "vfs.s3.connect_timeout_ms" : "3000"
   "vfs.s3.endpoint_override" : ""
   "vfs.s3.max_parallel_ops" : "8"
   "vfs.s3.multipart_part_size" : "5242880"
   "vfs.s3.proxy_host" : ""
   "vfs.s3.proxy_password" : ""
   "vfs.s3.proxy_port" : "0"
   "vfs.s3.proxy_scheme" : "https"
   "vfs.s3.proxy_username" : ""
   "vfs.s3.region" : "us-east-1"
   "vfs.s3.request_timeout_ms" : "3000"
   "vfs.s3.scheme" : "https"
   "vfs.s3.use_virtual_addressing" : "true"


TileDB allows you also to iterate only over the config parameters
with a certain *prefix* as follows:

.. content-tabs::

   .. tab-container:: cpp
      :title: C++

      .. code-block:: c++

        Config config;

        // Print only the S3 settings
        std::cout << "\nVFS S3 settings:\n";
        for (auto i = config.begin("vfs.s3."); i != config.end(); ++i) {
          auto& p = *i;
          std::cout << "\"" << p.first << "\" : \"" << p.second << "\"\n";
        }

The above produces the following output. Observe that the prefix
is *stripped* from the retrieved parameter names.

.. code-block:: bash

   VFS S3 settings:
   "connect_max_tries" : "5"
   "connect_scale_factor" : "25"
   "connect_timeout_ms" : "3000"
   "endpoint_override" : ""
   "max_parallel_ops" : "8"
   "multipart_part_size" : "5242880"
   "proxy_host" : ""
   "proxy_password" : ""
   "proxy_port" : "0"
   "proxy_scheme" : "https"
   "proxy_username" : ""
   "region" : "us-east-1"
   "request_timeout_ms" : "3000"
   "scheme" : "https"
   "use_virtual_addressing" : "true"

Saving/Loading config to/from file
----------------------------------

You can save the configuration parameters you used in your program
into a (local) text file, and subsequently load them from the
file into a new TileDB config if needed as follows:

.. content-tabs::

   .. tab-container:: cpp
      :title: C++

      .. code-block:: c++

        // Save to file
        Config config;
        config["sm.tile_cache_size"] = 0;
        config.save_to_file("tiledb_config.txt");

        // Load from file
        Config config_load("tiledb_config.txt");
        std::string tile_cache_size = config_load["sm.tile_cache_size"];
        std::cout << "\nTile cache size after loading from file: " << tile_cache_size
                  << "\n";

The above code creates a config object, changes the tile cache size to ``0``,
and saves the entire configuration into a file. Next, it creates a new
config loading the values from the created file. Running the program
produces the following output. Observe that the loaded tile cache size
value is ``0``, which is the value we altered prior to saving the config
to the file.

.. code-block:: bash

   Tile cache size after loading from file: 0

Inspecting the contents of the exported config file, we get the following:

.. code-block:: bash

  $ cat tiledb_config.txt
  sm.array_schema_cache_size 10000000
  sm.check_coord_dups true
  sm.dedup_coords false
  sm.enable_signal_handlers true
  sm.fragment_metadata_cache_size 10000000
  sm.num_async_threads 1
  sm.num_tbb_threads -1
  sm.tile_cache_size 0
  vfs.file.max_parallel_ops 8
  vfs.min_parallel_size 10485760
  vfs.num_threads 8
  vfs.s3.connect_max_tries 5
  vfs.s3.connect_scale_factor 25
  vfs.s3.connect_timeout_ms 3000
  vfs.s3.max_parallel_ops 8
  vfs.s3.multipart_part_size 5242880
  vfs.s3.proxy_port 0
  vfs.s3.proxy_scheme https
  vfs.s3.region us-east-1
  vfs.s3.request_timeout_ms 3000
  vfs.s3.scheme https
  vfs.s3.use_virtual_addressing true

Note that ``vfs.s3.proxy_host``, ``vfs.s3.proxy_username`` and
``vfs.s3.proxy_password`` are not exported for security purposes.

Summary of Parameters
---------------------

Below we provide a table with all the TileDB configuration parameters,
along with their description and default values.

.. table:: TileDB config parameters
    :widths: auto

    ======================================    ===================     ==================================================
    **Parameter**                             **Default Value**       **Description**
    --------------------------------------    -------------------     --------------------------------------------------
    ``"sm.array_schema_cache_size"``          ``"10000000"``          The array schema cache size in bytes.
    ``"sm.check_coord_dups"``                 ``"true"``              This is applicable only if ``sm.dedup_coords`` is
                                                                      ``false``. If ``true``, an error will be thrown if
                                                                      there are cells with duplicate coordinates during
                                                                      sparse array writes. If ``false`` and there are
                                                                      duplicates, the duplicates will be written without
                                                                      errors, but the TileDB behavior could be
                                                                      unpredictable.
    ``"sm.dedup_coords"``                     ``"false"``             If ``true``, cells with duplicate coordinates
                                                                      will be removed during sparse array writes. Note
                                                                      that ties during deduplication are broken
                                                                      arbitrarily.
    ``"sm.enable_signal_handlers"``           ``"true"``              Determines whether or not TileDB will install
                                                                      signal handlers.
    ``"sm.fragment_metadata_cache_size"``     ``"10000000"``          The fragment metadata cache size in bytes.
    ``"sm.num_async_threads"``                ``"1"``                 The number of threads allocated for async queries.
    ``"sm.num_tbb_threads"``                  ``"-1"``                The number of threads allocated for the TBB thread
                                                                      pool (if TBB is enabled). **Note:** this is a
                                                                      whole-program setting. Usually this should not be
                                                                      modified from the default. See also the
                                                                      documentation for TBB's ``task_scheduler_init``
                                                                      class.
    ``"sm.tile_cache_size"``                  ``"10000000"``          The tile cache size in bytes.
    ``"vfs.num_threads"``                     # of cores              The number of threads allocated for VFS
                                                                      operations (any backend), per VFS instance.
    ``"vfs.file.max_parallel_ops"``           ``vfs.num_threads``     The maximum number of parallel operations on
                                                                      objects with ``file:///`` URIs.
    ``"vfs.min_parallel_size"``               ``"10485760"``          The minimum number of bytes in a parallel VFS
                                                                      operation (except parallel S3 writes, which are
                                                                      controlled by ``vfs.s3.multipart_part_size``).
    ``"vfs.s3.connect_max_tries"``            ``"5"``                 The maximum tries for a connection. Any ``long``
                                                                      value is acceptable.
    ``"vfs.s3.connect_scale_factor"``         ``"25"``                The scale factor for exponential backoff when
                                                                      connecting to S3. Any ``long`` value is
                                                                      acceptable.
    ``"vfs.s3.connect_timeout_ms"``           ``"3000"``              The connection timeout in ms. Any ``long`` value
                                                                      is acceptable.
    ``"vfs.s3.endpoint_override"``            ``""``                  The S3 endpoint, if S3 is enabled.
    ``"vfs.s3.max_parallel_ops"``             ``vfs.num_threads``     The maximum number of S3 backend parallel
                                                                      operations.
    ``"vfs.s3.multipart_part_size"``          ``"5242880"``           The part size (in bytes) used in S3 multipart
                                                                      writes. Any ``uint64_t`` value is acceptable.
                                                                      **Note:** ``vfs.s3.multipart_part_size *
                                                                      vfs.s3.max_parallel_ops`` bytes will be buffered
                                                                      before issuing multipart uploads in parallel.
    ``"vfs.s3.proxy_host"``                   ``""``                  The S3 proxy host.
    ``"vfs.s3.proxy_password"``               ``""``                  The S3 proxy password.
    ``"vfs.s3.proxy_port"``                   ``"0"``                 The S3 proxy port.
    ``"vfs.s3.proxy_scheme"``                 ``"https"``             The S3 proxy scheme.
    ``"vfs.s3.proxy_username"``               ``""``                  The S3 proxy username.
    ``"vfs.s3.region"``                       ``"us-east-1"``         The S3 region.
    ``"vfs.s3.request_timeout_ms"``           ``"3000"``              The request timeout in ms. Any ``long`` value is
                                                                      acceptable.
    ``"vfs.s3.scheme"``                       ``"https"``             The S3 scheme.
    ``"vfs.s3.use_virtual_addressing"``       ``"true"``              Determines whether to use virtual addressing
                                                                      or not.
    ``"vfs.hdfs.kerb_ticket_cache_path"``     ``""``                  HDFS kerb ticket cache path.
    ``"vfs.hdfs.name_node_uri"``              ``""``                  Name node for HDFS.
    ``"vfs.hdfs.username"``                   ``""``                  HDFS username.
    ======================================    ===================     ==================================================




HISTORY.md
# In Progress

## New features

* Added string type support: ASCII, UTF-8, UTF-16, UTF-32, UCS-2, UCS-4 (PR #415)
* Added `TILEDB_ANY` datatype (PR #446)
* Added parallelized VFS read operations, enabled by default (PR #499)

## Improvements

* Added parallel S3, POSIX, and Win32 writes, enabled by default.
* Got rid of special S3 "directory objects"
* Refactored sparse reads, making them simpler and more amenable to parallelization.
* Refactored dense reads, making them simpler and more amenable to parallelization.
* Refactored dense ordered writes, making them simpler and more amenable to parallelization.
* Refactored unordered writes, making them simpler and more amenable to parallelization.
* Refactored global writes, making them simpler and more amenable to parallelization.
* Added ability to cancel pending background/async tasks. SIGINT signals now cancel pending tasks.
* Sparse read performance improvements with parallelism (using TBB as a dependency).
* Async queries now use a configurable number of background threads (default number of threads is 1).
* Added checks for duplicate coordinates and option for coordinate deduplication.
* Map usage via the C++ API `operator[]` is faster, similar to the `MapItem` path.

## Bug Fixes

* Fixed bugs with reads/writes of variable-sized attributes.
* Fixed file locking issue with simultaneous queries.
* Fixed S3 issues with simultaneous queries within the same context.

## API additions

### C API

* Added `tiledb_query_finalize` function. 
* Added `tiledb_vfs_get_config` function.
* Added `vfs.num_threads` and `vfs.min_parallel_size` config parameters.
* Added `vfs.{s3,file}.max_parallel_ops` config parameters.
* Added `vfs.s3.multipart_part_size` config parameter.
* Added `vfs.s3.proxy_{scheme,host,port,username,password}` config parameters.
* Added `tiledb_ctx_cancel_tasks` function.
* Added `sm.num_async_threads`, `sm.num_tbb_threads`, and `sm.enable_signal_handlers` config parameters.
* Added `tiledb_kv_has_key` to check if a key exists in the key-value store.
* Added `tiledb_kv_free`.
* Added `tiledb_array_{open, close, free}`.
* Added `tiledb_array_alloc`
* Added `tiledb_kv_alloc`
* Added `tiledb_kv_iter_alloc` which takes as input a kv object
* Added `tiledb_query_set_buffer` which sets a single attribute buffer
* Added `tiledb_query_get_type`
* Added `tiledb_array_get_query_type`
* Added `tiledb_kv_schema_{set,get}_capacity`.
* Added `tiledb_query_has_results`
* Added `tiledb_kv_is_dirty`
* Added `tiledb_kv_iter_reset`
* Added `tiledb_array_reopen`
* Added `tiledb_kv_reopen`
* Added `tiledb_array_max_buffer_size` and `tiledb_array_max_buffer_size_var`
* Added `tiledb_array_is_open`

### C++ API
* Support for trivially copyable objects, such as a custom data struct, was added. They will be backed by an `sizeof(T)` sized `char` attribute.
* `Attribute::create<T>` can now be used with compound `T`, such as `std::string` and `std::vector<T>`, and other
  objects such as a simple data struct.
* Added a `Dimension::create` factory function that does not take tile extent,
  which sets the tile extent to `NULL`.
* Added `Query::finalize()` function.
* Added `Context::cancel_tasks()` function.
* `tiledb::Attribute` can now be constructed with an enumerated type (e.x. `TILEDB_CHAR`).
* A `tiledb::Map` defined with only one attribute will allow implicit usage, e.x. `map[key] = val` instead of `map[key][attr] = val`.
* Added `Map::has_key` to check for key presence.
* `MapIter` can be used to create iterators for a map.
* Added `Array::{open, close}`
* Added `Map::{open, close}`
* Added `Array::query_type`
* Added `Query::query_type`
* Added optional attributes argument in `Map::Map` and `Map::open`
* Added `MapSchema::set_capacity` and `MapSchema::capacity`
* Added `Query::has_results`
* Added `Map::is_dirty`
* Added `MapIter::reset`
* Added an extra `Query` constructor that omits the query type (this is inherited from the input array).
* Changed the return type of the `Query` setters to return the object reference.
* Added `Array::reopen`
* Added `Map::reopen`
* Added `Stats` class (wraps C API `tiledb_stats_*` functions)
* Added `Config::save_to_file`
* Added `Array::is_open`

## Breaking changes

### C API

* `tiledb_query_finalize` must **always** be called before `tiledb_query_free` after global-order writes.
* Removed `tiledb_vfs_move` and added `tiledb_vfs_move_file` and `tiledb_vfs_move_dir` instead.
* Removed `force` argument from `tiledb_vfs_move_*` and `tiledb_object_move`.
* Removed `vfs.s3.file_buffer_size` config parameter.
* Removed `tiledb_query_get_attribute_status`.
* All `tiledb_*_free` functions now return `void` and do not take `ctx` as input (except for `tiledb_ctx_free`).
* Changed signature of `tiledb_kv_close` to take a `tiledb_kv_t*` argument instead of `tiledb_kv_t**`.
* Renamed `tiledb_domain_get_rank` to `tiledb_domain_get_ndim` to avoid confusion with matrix def of rank.
* Changed signature of `tiledb_array_get_non_empty_domain`.
* Removed `tiledb_array_compute_max_read_buffer_sizes`.
* Changed signature of `tiledb_{array,kv}_open`.
* Removed `tiledb_kv_iter_create`
* Renamed all C API functions that create TileDB objects from `tiledb_*_create` to `tiledb_*_alloc`.
* Removed `tiledb_query_set_buffers`
* Removed `tiledb_query_reset_buffers`
* Added query type argument to `tiledb_array_open`
* Changed argument order in `tiledb_config_iter_alloc`, `tiledb_ctx_alloc`, `tiledb_attribute_alloc`, `tiledb_dimension_alloc`, `tiledb_array_schema_alloc`, `tiledb_kv_schema_load`, `tiledb_kv_get_item`, `tiledb_vfs_alloc`

### C++ API
* Fixes with `Array::max_buffer_elements` and `Query::result_buffer_elements` to comply with the API docs. `pair.first` is the number of elements of the offsets buffer. If `pair.first` is 0, it is a fixed-sized attribute or coordinates.
* `std::array<T, N>` is backed by a `char` tiledb attribute since the size is not guaranteed.

examples/cpp_api/config.cc
/**
 * @file   config.cc
 *
 * @section LICENSE
 *
 * The MIT License
 *
 * @copyright Copyright (c) 2017-2018 TileDB, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @section DESCRIPTION
 *
 * This is a part of the TileDB quickstart tutorial:
 *   https://docs.tiledb.io/en/latest/config.html
 *
 * This program shows how to set/get the TileDB configuration parameters.
 */

#include <iostream>
#include <tiledb/tiledb>

using namespace tiledb;

void set_get_config_ctx_vfs() {
  // Create config objects
  Config config, config_ctx, config_vfs;

  // Set/Get config to/from ctx
  Context ctx(config);
  config_ctx = ctx.config();
  (void)config_ctx;

  // Set config to VFS
  VFS vfs(ctx, config);
  config_vfs = vfs.config();
  (void)config_vfs;
}

void set_get_config() {
  Config config;

  // Set value
  config["vfs.s3.connect_timeout_ms"] = 5000;

  // Append parameter segments with successive []
  config["vfs."]["s3."]["endpoint_override"] = "localhost:8888";

  // Get value
  std::string tile_cache_size = config["sm.tile_cache_size"];
  std::cout << "Tile cache size: " << tile_cache_size << "\n\n";
}

void print_default() {
  Config config;
  std::cout << "Default settings:\n";
  for (auto& p : config) {
    std::cout << "\"" << p.first << "\" : \"" << p.second << "\"\n";
  }
}

void iter_config_with_prefix() {
  Config config;

  // Print only the S3 settings
  std::cout << "\nVFS S3 settings:\n";
  for (auto i = config.begin("vfs.s3."); i != config.end(); ++i) {
    auto& p = *i;
    std::cout << "\"" << p.first << "\" : \"" << p.second << "\"\n";
  }
}

void save_load_config() {
  // Save to file
  Config config;
  config["sm.tile_cache_size"] = 0;
  config.save_to_file("tiledb_config.txt");

  // Load from file
  Config config_load("tiledb_config.txt");
  std::string tile_cache_size = config_load["sm.tile_cache_size"];
  std::cout << "\nTile cache size after loading from file: " << tile_cache_size
            << "\n";
}

int main() {
  set_get_config_ctx_vfs();
  set_get_config();
  print_default();
  iter_config_with_prefix();
  save_load_config();

  return 0;
}